scio-core/src/main/scala/com/spotify/scio/coders/Coder.scala

```diff
@@ -173,88 +165,86 @@ final private case class DisjunctionCoder[T, Id](
   override def getCoderArguments: JList[_ <: BCoder[_]] = coders.values.toList.asJava
 }
 
-final private[scio] case class LazyCoder[T](
+sealed private[scio] trait DelegatingCoder[T] extends BCoder[T] {
+
+  def bcoder: BCoder[T]
+
+  override def encode(value: T, os: OutputStream): Unit =
+    bcoder.encode(value, os)
+  override def encode(value: T, os: OutputStream, context: BCoder.Context): Unit =
+    bcoder.encode(value, os, context)
+  override def decode(is: InputStream): T =
+    bcoder.decode(is)
+  override def decode(is: InputStream, context: BCoder.Context): T =
+    bcoder.decode(is, context)
+  override def getCoderArguments: JList[_ <: BCoder[_]] =
+    bcoder.getCoderArguments
+  override def verifyDeterministic(): Unit =
+    bcoder.verifyDeterministic()
+  override def consistentWithEquals(): Boolean =
+    bcoder.consistentWithEquals()
+  override def structuralValue(value: T): AnyRef =
+    bcoder.structuralValue(value)
+  override def isRegisterByteSizeObserverCheap(value: T): Boolean =
+    bcoder.isRegisterByteSizeObserverCheap(value)
+  override def registerByteSizeObserver(value: T, observer: ElementByteSizeObserver): Unit =
+    bcoder.registerByteSizeObserver(value, observer)
+  override def getEncodedTypeDescriptor(): TypeDescriptor[T] =
+    bcoder.getEncodedTypeDescriptor
+}
+
+private[scio] case class RefCoder[T](var bcoder: BCoder[T]) extends DelegatingCoder[T]
+
+private[scio] object RefCoder {
+  def apply[T](): RefCoder[T] = new RefCoder[T](null)
+}
+
+final private[scio] class LazyCoder[T](
   typeName: String,
-  o: CoderMaterializer.CoderOptions
-)(coder: Coder[T])
```

Comment:
serializing this may not be deterministic as `coder` could have been derived by a macro, depending on the compiler order, some `anonfun$x` could change.
This also increases serialized graph size.